rclcpp_components/src/component_manager.cpp

```diff
@@ -0,0 +1,203 @@
+// Copyright 2019 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "component_manager.hpp"
+
+#include <functional>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "ament_index_cpp/get_resource.hpp"
+#include "rcpputils/filesystem_helper.hpp"
+#include "rcpputils/split.hpp"
+
+using namespace std::placeholders;
+
+namespace rclcpp_components
+{
+
+ComponentManager::ComponentManager(
+  std::weak_ptr<rclcpp::executor::Executor> executor)
+: Node("ComponentManager"),
+  executor_(executor)
+{
+  loadNode_srv_ = create_service<LoadNode>("~/_container/load_node",
+      std::bind(&ComponentManager::OnLoadNode, this, _1, _2, _3));
+  unloadNode_srv_ = create_service<UnloadNode>("~/_container/unload_node",
+      std::bind(&ComponentManager::OnUnloadNode, this, _1, _2, _3));
+  listNodes_srv_ = create_service<ListNodes>("~/_container/list_nodes",
+      std::bind(&ComponentManager::OnListNodes, this, _1, _2, _3));
+}
+
+void
+ComponentManager::OnLoadNode(
+  const std::shared_ptr<rmw_request_id_t> request_header,
+  const std::shared_ptr<LoadNode::Request> request,
+  std::shared_ptr<LoadNode::Response> response)
+{
+  (void) request_header;
+
+  std::string content;
+  std::string base_path;
+  if (!ament_index_cpp::get_resource(
+      "rclcpp_components", request->package_name, content, &base_path))
+  {
+    RCLCPP_ERROR(get_logger(), "Could not find requested resource in ament index");
+    response->error_message = "Could not find requested resource in ament index";
+    response->success = false;
+    return;
+  }
+
+  std::vector<std::string> lines = rcpputils::split(content, '\n', true);
+  for (const auto & line : lines) {
+    std::vector<std::string> parts = rcpputils::split(line, ';');
```

Comment:
It would be nice, I think, to see this kind of logic put into a function or static method, and then unit tested. Most of the other stuff in this pr is hard to test due to the nature of needing end-to-end shared libraries and ament index and such, but unit testing what we can would be nice to have.